stride/physics/iso_elastic/devito.py

```diff
@@ -0,0 +1,398 @@
+
+        # Set geometry and wavelet
+        wavelets = wavelets.data
+
+        self._src_scale = 1000
+        self._max_wavelet = np.max(np.abs(wavelets)) + 1e-31
+
+        window = scipy.signal.get_window(('tukey', 0.01), self.time.num, False)
+        window = window.reshape((self.time.num, 1))
+
+        self.dev_grid.vars.src.data[:] = wavelets.T * self._src_scale / self._max_wavelet * window
+
+        if self.interpolation_type == 'linear':
+            self.dev_grid.vars.src.coordinates.data[:] = shot.source_coordinates
+            self.dev_grid.vars.rec_tau.coordinates.data[:] = shot.receiver_coordinates
+
+    async def run_forward(self, wavelets, vp, vs, rho, **kwargs):
+        """
+        Run the state or forward problem.
+
+        Parameters
+        ----------
+        wavelets : Traces
+            Source wavelets.
+        vp : ScalarField
+            Compressional (acoustic) speed of sound of the medium, in [m/s].
+        vs : ScalarField
+            Transverse (shear) speed of sound of the medium, in [m/s].
+        rho : ScalarField
+            Density of the medium in [kg/m^3].
+        problem : Problem
+            Sub-problem being solved by the PDE.
+
+        Returns
+        -------
+
+        """
+        functions = dict(src=self.dev_grid.vars.src, )
+
+        self.state_operator.run(dt=self.time.step,
+                                **functions,
+                                **kwargs.pop('devito_args', {}))
+
+    async def after_forward(self, wavelets, vp, vs, rho, **kwargs):
+        """
+        Clean up after the state run and retrieve the time traces.
+
+        Parameters
+        ----------
+        wavelets : Traces
+            Source wavelets.
+        vp : ScalarField
+            Compressional (acoustic) speed of sound of the medium, in [m/s].
+        vs : ScalarField
+            Transverse (shear) speed of sound of the medium, in [m/s].
+        rho : ScalarField
+            Density of the medium in [kg/m^3].
+        problem : Problem
+            Sub-problem being solved by the PDE.
+
+        Returns
+        -------
+        Traces
+            Time traces produced by the state run.
+
+        """
+        problem = kwargs.pop('problem')
+        shot = problem.shot
+
+        self.wavefield = None
+
+        traces_data = np.asarray(self.dev_grid.vars.rec_tau.data, dtype=np.float32).T
+        traces_data *= self._max_wavelet / self._src_scale
+        traces = shot.observed.alike(name='modelled', data=traces_data)
+
+        self.dev_grid.deallocate('p')
+        self.dev_grid.deallocate('laplacian')
+        self.dev_grid.deallocate('src')
+        self.dev_grid.deallocate('rec')
+        self.dev_grid.deallocate('vp')
+        self.dev_grid.deallocate('vp2')
+        self.dev_grid.deallocate('inv_vp2')
+        self.dev_grid.deallocate('rho')
+        self.dev_grid.deallocate('buoy')
+        self.dev_grid.deallocate('alpha')
+        self.boundary.deallocate()
```

Comment:
you need to deallocate the stuff that you have allocated before, none of these things have been allocated

Reply:
Not sure what has been allocated where tbh

Reply:
Correction made